docs/buffers.md

```diff
@@ -0,0 +1,104 @@
+# Buffer Protocol and NumPy Arrays
+
+CSnakes supports the Python Buffer Protocol for `bytes` and `bytearray` types. The Buffer Protocol is a low-level interface for reading and writing raw bytes from Python objects.
+The `IPyBuffer` interface is used to represent Python objects that support the Buffer Protocol. It has methods for accessing the raw data of the buffer in a Read-Only or Read-Write Span.
+
+Since NumPy ndarrays also support the Buffer Protocol, you can use the `IPyBuffer` interface to efficiently read and write data from NumPy arrays.
+
+`typing.Buffer` (`collections.abc.Buffer`) was introduced in Python 3.12, but for older versions you can import `Buffer` from the `typing_extensions` package on PyPi. 
+
+For example:
+
+```python
+try:
+    from collections.abc import Buffer
+except ImportError:
+    from typing_extensions import Buffer
+
+import numpy as np
+
+def example_array() -> Buffer:
+    return np.array([True, False, True, False, False], dtype=np.bool_)
+```
+
+In this example, the `example_array` function returns a NumPy array of boolean values. The `Buffer` type hint indicates that the function returns an object that supports the Buffer Protocol.
+
+From C#, CSnakes will return an implementation of `CSnakes.Runtime.Python.IPyBuffer` that wraps the buffer and provides access to the data as a Span.
+
+For example:
+
+```csharp
+// Where testModule is a Python module that contains the example_array function
+var bufferObject = testModule.ExampleArray();
+
+// check if the buffer is scalar (single dimensional)
+if (bufferObject.IsScalar) {
+
+// Get the buffer contents as a Span of bool
+	Span<bool> result = bufferObject.AsBoolSpan();
+	Console.WriteLine(result[0]); // True
+	Console.WriteLine(result[4]); // False
+}
+```
+
+Since Span is writeable, you can also modify the buffer contents from C# and the changes will be reflected in the Python object.
```

Comment:
I feel like this should be more called out in big flashing red lights. The fact that it's writable, allowing you to write to memory you don't "own" is a very dangerous thing.

Maybe we should call out that it's recommended that you use the `ReadOnly` version unless you have a very explicit need.